sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -438,192 +423,162 @@ class GpuDynamicPartitionDataSingleWriter(
-      distinctKeys = distinctAndSort(partitionColumns)
-      val partitionIndexes = splitIndexes(partitionColumns, distinctKeys)
-      partitionColumns.close()
-      partitionColumns = null
+    val partDataTypes = description.partitionColumns.map(_.dataType).toArray
+    val outDataTypes = description.dataColumns.map(_.dataType).toArray
+    val (splits, cbKeys) = withResource(cb) { cb =>
+      val (partitionIndexes, cbKeys) = withResource(getPartitionColumns(cb)) { partitionColumns =>
+        withResource(distinctAndSort(partitionColumns)) { distinctKeys =>
+          val partitionIndexes = splitIndexes(partitionColumns, distinctKeys)
+          val cbKeys = copyToHostAsBatch(distinctKeys, partDataTypes)
+          (partitionIndexes, cbKeys)
+        }
+      }
 
       // split the original data on the indexes
-      outputColumns = getOutputColumns(cb)
-      val outDataTypes = description.dataColumns.map(_.dataType).toArray
-      cb.close()
-      needToCloseBatch = false
-      splits = outputColumns.contiguousSplit(partitionIndexes: _*)
-      outputColumns.close()
-      outputColumns = null
-
-      cbKeys = copyToHostAsBatch(distinctKeys, partDataTypes)
-      distinctKeys.close()
-      distinctKeys = null
-
-      // Use the existing code to convert each row into a path. It would be nice to do this on
-      // the GPU, but the data should be small and there are things we cannot easily support
-      // on the GPU right now
-      import scala.collection.JavaConverters._
-      val paths = cbKeys.rowIterator().asScala.map(getPartitionPath)
-
-      paths.toArray.zip(splits).foreach(combined => {
-        val table = combined._2.getTable
-        val batch = GpuColumnVector.from(table, outDataTypes)
-        val partPath = combined._1
-
-        // If fall back from for `GpuDynamicPartitionDataConcurrentWriter`, we should get the
-        // saved status
-        var savedStatus: Option[WriterStatusWithCaches] = None
-
-        if (currentPartPath != partPath) {
-          val previousPartPath = currentPartPath
-          currentPartPath = partPath
+      val splits = withResource(getOutputColumns(cb)) { outputColumns =>
+        outputColumns.contiguousSplit(partitionIndexes: _*)
+      }
+      (splits, cbKeys)
+    }
 
-          // see a new partition, close the old writer
-          val previousWriterStatus = currentWriterStatus
-          if (previousWriterStatus != null) {
-            releaseWriter(previousWriterStatus.outputWriter)
-          }
+    // Use the existing code to convert each row into a path. It would be nice to do this on
+    // the GPU, but the data should be small and there are things we cannot easily support
+    // on the GPU right now
+    import scala.collection.JavaConverters._
+    withResource(splits) { splits =>
+      withResource(cbKeys) { cbkeys =>
+        val paths = cbKeys.rowIterator().asScala.map(getPartitionPath)
+        paths.toArray.zip(splits).foreach( combined => {
```

Comment:
```suggestion
        paths.toArray.zip(splits).foreach{ case (path, split) =>
          .....
        }
```
Why does it need `toArray` here ?